lib/buffer.js

```diff
@@ -539,11 +539,29 @@ Buffer.prototype.inspect = function inspect() {
   return '<' + this.constructor.name + ' ' + str + '>';
 };
 
+function assertInteger(val, name) {
+  if (typeof val !== 'undefined' &&
+      (typeof val !== 'number' ||
+       +val !== (val >> 0))) {
+    throw new TypeError(`"${name}" argument must be an integer`);
+  }
+}
 
-Buffer.prototype.compare = function compare(b) {
+Buffer.prototype.compare = function compare(b, start, end, thisStart, thisEnd) {
   if (!(b instanceof Buffer))
     throw new TypeError('Argument must be a Buffer');
 
+  assertInteger(start, 'targetStart');
+  assertInteger(thisEnd, 'sourceEnd');
+  assertInteger(end, 'targetEnd');
+  assertInteger(thisStart, 'sourceStart');
```

Comment:
My above handles the specific case to always coerce ranges for `toString()` and `slice()` (which aren't allowed to throw). Though in the past I've tried to enforce more strict range checks. If we want to proceed down that path (not sure of this now that we're using typed arrays, which seem like they don't like to throw) then it should look something like (going to use new syntax. benchmark shows minor performance lose that should be mostly recovered in v5):
```js
Buffer.prototype.compare = function compare(b,
                                            start = 0,
                                            end = b.length,
                                            thisStart = 0,
                                            thisEnd = this.length) {
  if (thisStart >= thisEnd && start >= end)
    return 0;
  if (thisStart >= thisEnd)
    return -1;
  if (start >= end)
    return 1;

  if (start < 0 ||
      end > b.length ||
      thisStart < 0 ||
      thisEnd > this.length) {
    throw new RangeError('out of range index');
  }

  // Range comparisons above handle the case where end
  // and thisEnd are < 0. So can do uint32 coercion without
  // wrap around. Properly handles coercion of falsey/NaN
  // values.
  start >>>= 0;
  end >>>= 0;
  thisStart >>>= 0;
  thisEnd >>>= 0;

  return binding.compareOffset(this, b, start, thisStart, end, thisEnd);
}
```